### DISCUSSION TITLE
PixArt with dreambooth

### DISCUSSION THREAD

**Discussion by apluka34** [PixArt-alpha/PixArt-alpha#27] [Q&A] [answered]
Has anyone tried to train pixart alpha with dreambooth?

I have tried the scripts train_dreambooth.py from diffusers, but PixArt seems to have different architecture( for example transformer instead of unet for the denoising model). Are there any tutorials, thanks?

**Comment by lawrence-cj** [comment]
Will look into it soon. Bandwidth is full now.

**Comment by apluka34** [comment]
Hi @lawrence-cj, I have changed my code to adapt DreamBooth for Pixart Alpha but when I infer it doesn't generate the object I want. I ask if my code is somehow wrong? In the forward method of transformer, did I miss something?
Thank you!

![image](https://github.com/PixArt-alpha/PixArt-alpha/assets/79373948/e3590a9d-968a-4320-beda-25fddd4b600c)
![image](https://github.com/PixArt-alpha/PixArt-alpha/assets/79373948/03f06a71-f2dc-4f9e-bc82-c77eb2bea518)

**Reply by lawrence-cj** [reply]
Have you trained DreamBooth model by yourself?

**Reply by apluka34** [reply]
Yes, but it didn't generate the object I want. I trained with instance prompt "a photo of sks dog" and it generated with a random dog. Maybe there are problems with my training code scripts.

All I did is to change the code train_dreambooth.py from diffusers to adapt it with Pixart-Alpha-Pipeline instead of DiffusionPipeline, and I also used transformer for the denoising model (instead of unet). The code runs without any errors. That's why I wonder where did I go wrong?

I have read the paper, and there are results of training DreamBooth with Pixart Alpha, can you give me some tutorials or training scripts examples? Thank you so much.

Here is my training scripts:
https://pastecode.io/s/hrofzbuc

**Reply by lawrence-cj** [reply]
Actullay, we train the DreamBooth with our github codebase instead the diffusers one. We will release DreamBooth releted code in the very near future.

**Reply by apluka34** [reply]
I am looking forward to seeing this scripts. Thanks @lawrence-cj

**Comment by lawrence-cj** [comment, marked as answer]
The Dreambooth [training scripts](https://github.com/PixArt-alpha/PixArt-alpha/blob/master/scripts/train_dreambooth.py) are released. Feel free to test them Plz.

**Reply by storuky** [reply]
Hey @lawrence-cj Thank you for you work!👏🏻
First, A fix to Accelerator calling: [logging_dir](https://github.com/PixArt-alpha/PixArt-alpha/blob/156452504383a3bcb768219ef2d734a15dd1b655/scripts/train_dreambooth.py#L210C12-L210C12) is deprecated since 0.18 version of accelerate, so lets change it to project_dir instead.
Should I open this minor PR or you can change it on your side?

Second, it raises an error in 54 line of diffusion/utils/checkpoint.py:
```
KeyError: 'state_dict_ema'
```
As far as it's just del so does basic condition solve this?
```
if 'state_dict_ema' in checkpoint:
  del checkpoint['state_dict_ema']['pos_embed']
```
and some similar conditions below.

I'm using [PixArt-XL-2-1024-MS.pth](https://huggingface.co/PixArt-alpha/PixArt-alpha/blob/main/PixArt-XL-2-1024-MS.pth) from HF. Any thoughts on that?
Thank you!

**Reply by storuky** [reply]
Ok, extra conditions solved this issue.
Next: [in this line](https://github.com/PixArt-alpha/PixArt-alpha/blob/156452504383a3bcb768219ef2d734a15dd1b655/scripts/train_dreambooth.py#L60) cache_dir is looking for a t5-v1_1-xxl inside provided dir. So a workaround is to pass a root path to PixArt-alpha models folder and not to t5-v1_1-xxl which is weird.

**Reply by storuky** [reply]
Now I'm facing the issue `TypeError: 'NoneType' object is not callable` in [this line](https://github.com/PixArt-alpha/PixArt-alpha/blob/156452504383a3bcb768219ef2d734a15dd1b655/scripts/train_dreambooth.py#L69C4-L69C4).

**Reply by lawrence-cj** [reply]
1. It would be nice if you can throw a PR for me next time. I'll change the logging_dir ---> project_dir.
2. Solved the ckpt issue.
3. As for `t5-v1_1-xxl `, We just keep it the same as DeepFloyd base code here: 
https://github.com/deep-floyd/IF/blob/ffc8163891682beaca0c5eb6b9077860a3bc6509/deepfloyd_if/modules/t5.py#L19
4. For the NoneType issue, have you pulled all the commits to your branch? May I have more information about the error?

**Reply by storuky** [reply]
@lawrence-cj yeah, I'm in a master branch. I have multiple tries and cloned repo all the time.
Added pdb.set_trace() before `for epoch in range(start_epoch + 1, config.num_epochs + 1):`:
```
(Pdb) train_dataloader
<accelerate.data_loader.DataLoaderShard object at 0x7fe12bbf7c70>

(Pdb) dir(train_dataloader)
['_DataLoader__initialized', '_DataLoader__multiprocessing_context', '_IterableDataset_len_called', '__annotations__', '__class__', '__class_getitem__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__orig_bases__', '__parameters__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_auto_collation', '_dataset_kind', '_drop_last', '_get_iterator', '_index_sampler', '_is_accelerate_prepared', '_is_protocol', '_iterator', 'batch_sampler', 'batch_size', 'begin', 'check_worker_number_rationality', 'collate_fn', 'dataset', 'device', 'drop_last', 'end', 'end_of_dataloader', 'generator', 'gradient_state', 'iteration', 'multiprocessing_context', 'num_workers', 'persistent_workers', 'pin_memory', 'pin_memory_device', 'prefetch_factor', 'remainder', 'reset', 'rng_types', 'sampler', 'set_epoch', 'skip_batches', 'synchronized_generator', 'timeout', 'total_batch_size', 'total_dataset_length', 'worker_init_fn']

(Pdb) train_dataloader.dataset
<diffusion.data.datasets.Dreambooth.DreamBooth object at 0x7fe16846b970>

(Pdb) len(train_dataloader.dataset)
5

(Pdb) train_dataloader.num_workers
1

(Pdb) train_dataloader.__iter__()
<generator object DataLoaderShard.__iter__ at 0x7fe192997740>

(Pdb) for step, batch in enumerate(train_dataloader): print(step)
*** TypeError: 'NoneType' object is not callable
```
I played in interactive shell with train_dataloader. It has __iter__ and it's DataLoaderShard instance. But when running a loop over enumerate I'm getting `*** TypeError: 'NoneType' object is not callable`

I can execute any command you want in Pdb to help with debug!

**Reply by storuky** [reply]
This is a trace I have on my machine:
```
Traceback (most recent call last):
  File "/home/storuky/ml/pix_train/PixArt-alpha/scripts/train_dreambooth.py", line 308, in <module>
    train()
  File "/home/storuky/ml/pix_train/PixArt-alpha/scripts/train_dreambooth.py", line 69, in train
    for step, batch in enumerate(train_dataloader):
  File "/home/storuky/miniconda3/envs/pixart/lib/python3.9/site-packages/accelerate/data_loader.py", line 447, in __iter__
    self.set_epoch(self.iteration)
  File "/home/storuky/miniconda3/envs/pixart/lib/python3.9/site-packages/accelerate/data_loader.py", line 484, in set_epoch
    self.dataset.set_epoch(epoch)
TypeError: 'NoneType' object is not callable
```

**Reply by lawrence-cj** [reply]
What about replace code here: https://github.com/PixArt-alpha/PixArt-alpha/blob/d63a615aa6fe2a7e74cd9a1c2e9bc92c91c045a9/diffusion/data/datasets/Dreambooth.py#L74
with:
```
  def __getattr__(self, attr):
      try:
          return super().__getattr__(attr)
      except AttributeError:
          raise AttributeError(f"'{type(self).__name__}' object has no attribute '{attr}'")
```
or 
```
  def __getattr__(self, name):
      if name == "set_epoch":
          return lambda epoch: None
      raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
```

**Reply by storuky** [reply]
@lawrence-cj aha, it went forward and stucked here:
```
Traceback (most recent call last):
  File "/home/storuky/ml/pix_train/PixArt-alpha/scripts/train_dreambooth.py", line 309, in <module>
    train()
  File "/home/storuky/ml/pix_train/PixArt-alpha/scripts/train_dreambooth.py", line 109, in train
    f"epoch_eta:{eta_epoch}, time_all:{t:.3f}, time_data:{t_d:.3f}, lr:{lr:.3e}, s:({model.module.h}, {model.module.w}), "
  File "/home/storuky/miniconda3/envs/pixart/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1614, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(
AttributeError: 'PixArtMS' object has no attribute 'module'
```

**Reply by storuky** [reply]
Commenting it out seems to fixed this issue. But no logging anymore 😁

**Reply by storuky** [reply]
Removing module mentions in this logging line seems to fix it in a better way 😁

**Reply by lawrence-cj** [reply]
Maybe you ain't using DDP training? 
`s:({model.module.h}, {model.module.w})`
↓
`s:({model.h}, {model.w})`

**Reply by storuky** [reply]
And yeah, everything seems to be working. 🙏🏻
Next question will be how to convert result pth model to diffusers 🤔

**Reply by lawrence-cj** [reply]
https://github.com/PixArt-alpha/PixArt-alpha/blob/master/tools/convert_pixart_alpha_to_diffusers.py

**Reply by storuky** [reply]
@lawrence-cj cool! Could you make `only_transformer` param not required as far as it has default True value?

**Reply by lawrence-cj** [reply]
We would keep it the same the official one: https://github.com/huggingface/diffusers/blob/4e54dfe985293df9e6e86828d7a2763d076879f5/scripts/convert_pixart_alpha_to_diffusers.py#L195

**Reply by storuky** [reply]
Thank you so much!  Great job 👏🏻
Everything works smoothly!

**Reply by lawrence-cj** [reply]
Feel free to upload diffusers checkpoint to hugging Face. That would be great helpful. : )

**Reply by storuky** [reply]
@lawrence-cj  Actually, I'm not sure Finetuned models could be interesting for a community without clear Loras support. But anyway this is a good jumpstart for a community! 🎉
Btw, it's more flexible to fetch prompt from caption files instead of having them statically inside config_file. Do you have some plans to bring this functionality?

**Reply by lawrence-cj** [reply]
Of course. Looking forward to your PR. We look forward to everyone's contributions to make repo better.

**Comment by axel578** [comment]
I get this error for training, at the beginning it seems to train okay, even thow the loss (even thow when calculating cumulated loss on 1 epoch is still weirdly oscillating abnormally like not going down at all)

This issue happen sometimes randomly
`
2023-11-24 20:58:39,971 - PixArt - INFO - Steps [587][13/14]:total_eta: 2:05:29, epoch_eta:0:00:03, time_all:2.478, time_data:0.001, lr:5.000e-06loss:0.0003, grad_norm:0.0112
2023-11-24 20:58:43,115 - PixArt - INFO - Steps [588][14/14]:total_eta: 2:05:25, epoch_eta:0:00:00, time_all:3.143, time_data:0.511, lr:5.000e-06loss:0.1477, grad_norm:0.1474
Traceback (most recent call last):
  File "K:\pixaar\ppbx\lib\site-packages\torch\utils\data\dataloader.py", line 1132, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "C:\Users\me\AppData\Local\Programs\Python\Python310\lib\queue.py", line 179, in get
    raise Empty
_queue.Empty

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "K:\pixaar\PixArt-alpha\scripts\train_dreambooth.py", line 340, in <module>
    train()
  File "K:\pixaar\PixArt-alpha\scripts\train_dreambooth.py", line 101, in train
    for step, batch in enumerate(train_dataloader):
  File "K:\pixaar\ppbx\lib\site-packages\accelerate\data_loader.py", line 451, in __iter__
    current_batch = next(dataloader_iter)
  File "K:\pixaar\ppbx\lib\site-packages\torch\utils\data\dataloader.py", line 633, in __next__
    data = self._next_data()
  File "K:\pixaar\ppbx\lib\site-packages\torch\utils\data\dataloader.py", line 1328, in _next_data
    idx, data = self._get_data()
  File "K:\pixaar\ppbx\lib\site-packages\torch\utils\data\dataloader.py", line 1284, in _get_data
    success, data = self._try_get_data()
  File "K:\pixaar\ppbx\lib\site-packages\torch\utils\data\dataloader.py", line 1145, in _try_get_data
    raise RuntimeError('DataLoader worker (pid(s) {}) exited unexpectedly'.format(pids_str)) from e
RuntimeError: DataLoader worker (pid(s) 8560) exited unexpectedly
`

**Reply by lawrence-cj** [reply]
It seems you are training on a Windows environment?

**Reply by axel578** [reply]
yeah it is, but I was able to train for quite  a bit so it should work no ? I think its more of a code issue, I can go on linux thow but It would require me to install dual boot ( and wsl  2 isn't that much stable, sooooo much issues with wsl)

**Reply by lawrence-cj** [reply]
Actually, we run the code on Linux smoothly. Sry for that we don't have a windows system to debug. We highly recommend to run on Linux.
